apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RemoveMe/ResumableUpload.cs

```diff
@@ -0,0 +1,1077 @@
+﻿/*
+Copyright 2012 Google Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using System;
+using System.Collections;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Reflection;
+
+using Google.Apis.Http;
+using Google.Apis.Logging;
+using Google.Apis.Media;
+using Google.Apis.Requests;
+using Google.Apis.Services;
+using Google.Apis.Testing;
+using Google.Apis.Util;
+using Google.Apis.Json;
+
+namespace Google.Apis.Upload
+{
+    public sealed class ResumableUploadSessionOptions
+    {
+        public HttpClient HttpClient { get; set; }
+
+        internal ConfigurableHttpClient ConfigurableHttpClient { get { return HttpClient as ConfigurableHttpClient; } }
+
+        public Action<HttpRequestMessage> ModifySessionUriRequest { get; set; }
+
+        public ISerializer Serializer { get; set; }
+
+        public string ServiceName { get; set; }
+    }
+
+    public class TmpResumableUpload
+    {
+        protected ConfigurableHttpClient HttpClient { get; }
+
+        private ResumableUploadSessionOptions Options { get; }
+
+        private string SignedUrl { get; set; }
```

Comment:
I'm nervous about signed URLs ending up in the support library.

Is there actually any difference between a signed URL and any other session URI here? It's entirely possible that I'm missing something, but it sounds like there are two operations required to perform an upload:

- Create a session
- Upload the data

At the moment (currently released code) those are tied together. Why not have an *abstract* GetSessionUriAsync method, and then two subclasses: one which makes the request, and one which is prepopulated with a session URI. That session URI could have come from anywhere - it doesn't have to be via a signed URL.